### PR TITLE
Increase speed and parallelism of the limit algorithm and implement descending sorting

### PR DESCRIPTION
**Pull request by nils-braun**
This PR introduces four changes:
* it simplifies the logic for getting a piece of the dask dataframe starting at OFFSET and ending at LIMIT.
* this allows to not needing to calculate the dask dataframe before, but really only at the actual calculation. The calculated used to be needed for getting the partition boundaries, but they can now also only be calculated "lazily"
* as there is no need to do a full recalculation, I introduced a quick shortcut: if the first partition is already enough, just return this. This is a very typical use case when you just do a "LIMIT 10".
* it allows for descending sorting also in the first column

### CODE REVIEW

File: dask_sql/physical/rel/logical/sort.py [hunk 1]
```diff
@@ -79,6 +73,12 @@ def _apply_sort(
             raise ValueError("Can not sort a column with NaNs")
 
         df = df.set_index(first_sort_column, drop=False).reset_index(drop=True)
+        if not first_sort_ascending:
+            # As set_index().reset_index() always sorts ascending, we need to reverse
+            # the order inside all partitions and the order of the partitions itself
+            df = dd.from_delayed(
+                reversed([self.reverse_partition(p) for p in df.to_delayed()])
+            )
```
Comment (mrocklin) [inline dask_sql/physical/rel/logical/sort.py:81]: You may want to become familiar with the `df.partitions` iterable. 

```python
dd.concat(df.partitions[::-1])
```

Going between delayed and dataframes introduces some inconvenience in graph-handling.

File: dask_sql/physical/rel/logical/sort.py [hunk 2]
```diff
@@ -104,24 +103,32 @@ def _apply_offset(self, df: dd.DataFrame, offset: int, end: int) -> dd.DataFrame
...
         """
-        # As we need to calculate the partition size, we better persist
-        # the df. I think...
-        # TODO: check if this is the best thing to do
-        df = df.persist()
+        if not offset:
+            # We do a (hopefully) very quick check: if the first partition
+            # is already enough, we will just ust this
+            first_partition_length = df.map_partitions(lambda x: len(x)).to_delayed()[0]
+            first_partition_length = first_partition_length.compute()
```
Comment (mrocklin) [inline dask_sql/physical/rel/logical/sort.py:110]:
```python
len(df.partitions[0])
```

File: dask_sql/physical/rel/logical/sort.py (hunk 1 above)
Comment (nils-braun) [reply dask_sql/physical/rel/logical/sort.py:81]: Uh nice! That looks awesome. Yes, this is definitely much better.

Comment (mrocklin) [reply dask_sql/physical/rel/logical/sort.py:81]: At some point I think that we'll want to go through and remove all of the `to/from_delayed` calls.  This isn't critical, but it's nice to do from a performance standpoint (happy to go into this in more depth if you like).

Comment (nils-braun) [reply dask_sql/physical/rel/logical/sort.py:81]: Thanks for your advice, I would actually be really interested in your thoughts about [this part](https://github.com/nils-braun/dask-sql/pull/75/files#diff-1fcfb9a4ff196f6a0e8fb3847cc3d2300b9b3a91af77f0555e6a61a179757802R153). 

Probably you see it right away, but what I would like to achieve is basically two things
1. call a function on every partition, while also having the partition index as an input argument
2. calculate `partition_borders`  only when executing the DAG - not already when creating it (this is what I have done before this PR, and it slows down the process).

So far, the `delayed` function was all I could come up with, but I definitely think there is a better way.

Comment (nils-braun) [reply dask_sql/physical/rel/logical/sort.py]: I have created an issue to fix this, so we can go on with this PR.

**Pull request merged**